### PR TITLE
Fix for #39

### LINKED ISSUE #39: Synchronous interface
I wonder if there was any need for a synchronous interface to node-tmp. I for my part definitely want such an interface :smile:

Please see the follow up PR for an initial stab at such an alternate interface that introduces synchronous versions of the existing API calls.

The PR still has some unresolved issues (failing tests, name clashes), but before continuing implementing the required test cases I wanted to know whether the API and its implementation is something you would like to have in node-tmp.

### PR DESCRIPTION
**Pull request by silkentrance**
- fixes #39:synchronous interface

Please note that this is **work in progress** and presented here for review only.

I needed to partly fix the issue where certain tmp dirs and tmp files were not being removed (using name tests) in order to make this work.

This patch is almost ready, please review.

### CODE REVIEW

**Comment by silkentrance** [comment]
The build failed due to some unrelated issues with npm.

File: lib/tmp.js [hunk 1]
```diff
@@ -148,6 +155,29 @@ function _getTmpName(options, callback) {
   }());
 }
 
+function _getTmpNameSync(options) {
+  var
+    args = _parseArguments(options),
+    opts = args[0],
+    tries = opts.tries || 3;
+
+  if (isNaN(tries) || tries < 0)
```
Comment (raszi) [inline lib/tmp.js]: This should be generalized because it a code duplication.

File: lib/tmp.js [hunk 2]
```diff
@@ -250,12 +307,14 @@ function _createTmpDir(options, callback) {
     fs.mkdir(name, opts.mode || 0700, function _dirCreated(err) {
       if (err) return cb(err);
 
-      var removeCallback = _prepareRemoveCallback(
-        opts.unsafeCleanup
-          ? _rmdirRecursiveSync
-          : fs.rmdirSync.bind(fs),
-        name
-      );
+      var removeCallback = _prepareRemoveCallback(function (name) {
```
Comment (raszi) [inline lib/tmp.js]: Why do you need to wrap this into a function?

File: lib/tmp.js [hunk 3]
```diff
@@ -266,6 +325,32 @@ function _createTmpDir(options, callback) {
...
 }
 
+function _createTmpDirSync(options) {
+  var
+    args = _parseArguments(options),
+    opts = args[0];
+
+  var name = _getTmpNameSync(opts);
+  fs.mkdirSync(name, opts.mode || 0700);
+  var removeCallback = _prepareRemoveCallback(function (name) {
```
Comment (raszi) [inline lib/tmp.js]: This is also a duplicated code.

**Comment by raszi** [comment]
Thanks for the hard work @silkentrance

**Comment by silkentrance** [comment]
You're welcome. I thank you for providing this library!

Rebased to master.

**Comment by silkentrance** [comment]
@raszi Sorry some of the line comments got lost due to the --force commit.

I managed to refactor the removeCallback factories, will try to refactor existing code duplicates as well.

**Comment by silkentrance** [comment]
Build fails for older node versions due to missing or incompatible packages, e.g.

node@0.8: Error: No compatible version found: minimatch@'^1.0.0'

**Comment by silkentrance** [comment]
@raszi reverted back to vows@0.7.0 and implemented the failing tests in a different way, catch/returning the exception instead of just throwing it.

**Comment by silkentrance** [comment]
@raszi Refactoring _getTmpName and _getTmpNameSync in order to get rid of the duplicate code is not only difficult as the code is actually very different as one uses the callback and the other does not, but it will add to the complexity of the library for getting rid of the actual duplicate line

```
     tries = opts.tries || 3;
```

Instead I moved the shared constant '3' to a named constant DEFAULT_TRIES.

File: lib/tmp.js [hunk 4]
```diff
@@ -124,40 +158,16 @@ function _getTmpName(options, callback) {
...
-    template = opts.template,
-    templateDefined = !_isUndefined(template),
-    tries = opts.tries || 3;
+    tries = opts.tries || DEFAULT_TRIES;
 
   if (isNaN(tries) || tries < 0)
     return cb(new Error('Invalid tries'));
 
-  if (templateDefined && !template.match(TEMPLATE_PATTERN))
+  if (opts.template && !opts.template.match(/XXXXXX/))
```
Comment (raszi) [inline lib/tmp.js]: Please keep the newly introduced `TEMPLATE_PATTERN`.

File: lib/tmp.js [hunk 5]
```diff
@@ -173,6 +183,50 @@ function _getTmpName(options, callback) {
...
+function _getTmpNameSync(options) {
+  var
+    args = _parseArguments(options),
+    opts = args[0],
+    tries = opts.tries || DEFAULT_TRIES;
+
+  if (isNaN(tries) || tries < 0)
+    throw new Error('Invalid tries');
+
+  if (opts.template && !opts.template.match(/XXXXXX/))
```
Comment (raszi) [inline lib/tmp.js]: Same here.

**Comment by silkentrance** [comment]
Added missing comments and cleaned up a bit.

**Comment by silkentrance** [comment]
Fixed the issue with TEMPLATE_PATTERN.

**Comment by silkentrance** [comment]
New feature still needs to be documented in the readme.

**Comment by silkentrance** [comment]
Rebased to master.

**Comment by silkentrance** [comment]
Removed reference to vows@0.8.1 from commit message as it is no longer being required.

**Comment by silkentrance** [comment]
@raszi Updated readme, but you might want to have a look at it first...

**Comment by silkentrance** [comment]
@raszi Please provide some more feedback...

**Comment by silkentrance** [comment]
@raszi what about this sync interface, are you willing to merge this or not? still waiting for feedback, though.

**Comment by raszi** [comment]
I'll fix the issues in master.

**Comment by silkentrance** [comment]
@raszi Thanks for merging.

**Pull request merged**